near-sdk/src/store/vec/mod.rs

```diff
@@ -61,8 +115,8 @@ pub struct Vector<T>
 where
     T: BorshSerialize,
 {
-    len: u32,
-    values: IndexMap<T>,
+    pub(crate) len: u32,
+    pub(crate) values: IndexMap<T>,
 }
 
 //? Manual implementations needed only because borsh derive is leaking field types
```

Comment:
I mean, the main reason is to avoid loading all elements into memory when deserializing state. But yeah, it seems worth it to document that more specifically to avoid returning one of these structures from a function.

Ideally, we wouldn't even have this implementation, but the SDK codegen assumes that the state implements `Borsh(De)Serialize` to store and load state from storage

Reply:
>This is a bit surprising, worth documenting perhaps.

On the NEAR Con, one person was struggling with exactly they had a code like 

```rust
#[derive(borsh)]
struct Thing {
   components: Vec<Component>
}

#[near_bindgen]
impl Contract {
    pub fn call_me(&self, thing: Thing)
}
```

They also have trouble figuring out how to make that use borsh or json, and ingeneral got entangled in two different serialization formats.

Reply:
I mean, that is very unrelated to this PR. Perhaps worth improving docs/visibility though. @itegulov wdyt?

Reply:
Yeah, during NEARCon I have also seen a person who was confused about collections' (de)serialization semantics. Trying to generate an ABI actually helps to catch these kinds of errors as there is no `JsonSchema`/`BorshSchema` derivation for collections. We could perhaps somehow tailor the errors to communicate the reasoning (i.e. you are not supposed to use `Vec` for your public API) to the users.